### PR TITLE
PHOENIX-6907 Explain Plan should output region locations with servers

### PR DESCRIPTION
**Pull request by virajjasani**
Query syntax: `EXPLAIN WITH REGIONS {query}`

Example:

Assuming the scan is prepared for 4 region locations,

When `phoenix.max.region.locations.size.explain.plan` is set as `2`, only 2 region locations are printed:
```
CLIENT PARALLEL 4-WAY RANGE SCAN OVER N000001:N000002 [1,'a'] - [1,'b']
    SERVER MERGE [0.K3]
    SERVER FILTER BY FIRST KEY ONLY
 (region locations = [region=N000001:N000002,,1683068961616.4773aca00009beabd2316e0be5828ffa., hostname=localhost,55546,1683068927671, seqNum=9, region=N000001:N000002,e\x00\x00\x00\x00\x00\x00\x00\x00\x00,1683068961616.77243f1c70242067f53018949e146393., hostname=localhost,55546,1683068927671, seqNum=9]...total size = 4) 
CLIENT MERGE SORT
```

With default value of `phoenix.max.region.locations.size.explain.plan` as `5`, all region locations are printed:
```
CLIENT PARALLEL 4-WAY RANGE SCAN OVER N000001:N000002 [1,'a'] - [1,'b']
    SERVER MERGE [0.K3]
    SERVER FILTER BY FIRST KEY ONLY
 (region locations = [region=N000001:N000002,,1683068843354.6f20f10367637072dabf805a38ac8d55., hostname=localhost,55344,1683068805683, seqNum=9, region=N000001:N000002,e\x00\x00\x00\x00\x00\x00\x00\x00\x00,1683068843354.fdfd4eb63d0fff6d80da1d08726bfeb0., hostname=localhost,55344,1683068805683, seqNum=9, region=N000001:N000002,i\x00\x00\x00\x00\x00\x00\x00\x00\x00,1683068843354.effe18db5f4a093d005706910436ed18., hostname=localhost,55344,1683068805683, seqNum=9, region=N000001:N000002,o\x00\x00\x00\x00\x00\x00\x00\x00\x00,1683068843354.d7d0cec570fe9c626d39a0b2075918d2., hostname=localhost,55344,1683068805683, seqNum=9]) 
CLIENT MERGE SORT
```

### CODE REVIEW

File: phoenix-core/src/main/antlr3/PhoenixSQL.g [hunk 1]
```diff
@@ -452,7 +454,14 @@ oneStatement returns [BindableStatement ret]
 finally{ contextStack.pop(); }
     
 explain_node returns [BindableStatement ret]
-    :   EXPLAIN q=oneStatement {$ret=factory.explain(q);}
+    :   EXPLAIN (w=WITH)? (r=REGIONS)? q=oneStatement
```
Comment (jpisaac) [inline phoenix-core/src/main/antlr3/PhoenixSQL.g:457]: Would it make sense to not change what the EXPLAIN statement generates. But have all plan info/attributes generated all the time.

And have different kinds of plan output generators. 
We may want to look at how other DBs have implemented it.

Comment (virajjasani) [reply phoenix-core/src/main/antlr3/PhoenixSQL.g:457]: Sounds good, let me try this tomorrow i.e. let the source generate all info in the explain plan including region info, but it's upto `ExecutableExplainStatement` to control what to print.

**Comment by virajjasani** [comment]
@jpisaac addressed review comments

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java [hunk 2]
```diff
@@ -46,27 +61,37 @@
...
+    public static final String REGION_LOCATIONS = " (region locations = ";
+
     protected final StatementContext context;
     protected final TableRef tableRef;
     protected final GroupBy groupBy;
     protected final OrderBy orderBy;
     protected final HintNode hint;
     protected final Integer limit;
     protected final Integer offset;
+    private final ExecutorService executorService;
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:94]: Why do we need an executor and async processing?

**Review by jpisaac — COMMENTED** [review]
Mostly looks good, but not sure about the async calls?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java (hunk 2 above)
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:94]: this is because we want to have our own timeouts rather than relying on hbase client timeouts. 
if for some reason, we can't get region location due to some transient issue, client will keep retrying based on retry configs, but it would not be appropriate to block output of explain plan just because of that.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java [hunk 3]
```diff
@@ -284,6 +367,102 @@ protected void explain(String prefix, List<String> planSteps,
...
+                                    List<List<Scan>> scansList) {
+        Future<String> task = executorService.submit(
+                () -> getRegionLocationsForExplainPlan(explainPlanAttributesBuilder, scansList));
+        try {
+            String regionLocations = task.get(5, TimeUnit.SECONDS);
+            planSteps.add(regionLocations);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            LOGGER.error("Unable to retrieve region locations task result in 5s", e);
+            task.cancel(true);
+        }
```
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:388]: this would be our own timeout, and if we can't get region locations on time, we rather cancel the task than letting hbase client retry for very long time.

@jpisaac does this sound good to you?

Comment (jpisaac) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:388]: Creating an executor service for every query, might not be the right thing.

**Comment by virajjasani** [comment]
latest build results: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1598/13/PR_20General_20Check_20Report/

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java [hunk 4]
```diff
@@ -284,6 +355,106 @@ protected void explain(String prefix, List<String> planSteps,
...
+                return false;
+            }
+            RegionBoundary that = (RegionBoundary) o;
+            return Bytes.compareTo(startKey, that.startKey) == 0
+                    && Bytes.compareTo(endKey, that.endKey) == 0;
+        }
+
+        @Override
+        public int hashCode() {
+            return 0;
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:454]: why hashcode of zero?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java [hunk 5]
```diff
@@ -111,8 +126,63 @@ private String explainSkipScan() {
...
+                    regionLocations.add(regionLocation);
+                    regionBoundaries.add(regionBoundary);
+                }
+                currentKey = regionLocation.getRegion().getEndKey();
+                // condition1 = currentKey != END_ROW_KEY
+                // condition2 = endKeyIsEndOfTable == true
+                // condition3 = currentKey < endKey
+                // condition4 = includeEndKey == true
+                // condition5 = currentKey == endKey
+                // while (condition1 && (condition2 || condition3 || (condition4 && condition5)))
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:175]: nit: Might be easier on the reader if it is in plain layman's language

**Review by jpisaac — COMMENTED** [review]
Done with the review.Thanks

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java (hunk 4 above)
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:454]: since byte[] can't be compared for hashcode (as they don't have `hashcode()` impl), i had to use unique value.
if we were doing string based comparison, then i would have used string's hashcode and equals implementations directly.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java (hunk 5 above)
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java:178]: done, before do/while loop

**Review by jpisaac — APPROVED** [review]
LGTM +1 Thanks @virajjasani
